openmp/libomptarget/plugins-nextgen/amdgpu/src/rtl.cpp

```diff
@@ -2728,6 +2780,34 @@ struct AMDGPUDeviceTy : public GenericDeviceTy, AMDGenericDeviceTy {
     return Err;
   }
 
+  /// Detect if current architecture is an APU.
+  Error checkIfAPU() {
```

Comment:
I was talking with SuiChi about checking this. He says that HSA_PROFILE_FULL should be set if the pointters are usable https://alt-doc-doc-test.readthedocs.io/en/latest/ROCm_API_References/ROCr-Runtime.html#group__agentinfo_1gacafd4247e2a04cbe0ac0b3998c127532. This is apparently how it's used in rocm-info https://github.com/ROCm/ROCR-Runtime/blob/master/src/core/runtime/amd_gpu_agent.cpp#L115. So you need to query agent info with the profile as an argument.

Reply:
Host-allocated ppointers are usable on any gpu that supports unified memory, including MI200 and others.
This is not what we are determining here. What we are finding is whether the architecture is an APU, not a discrete GPU with unified memory capability. If this is changed to return true on MI200, it will make automatic zero-copy the default behavior on that gpu if XNACK is enabled. That's not what we want. What we want is to enable automatic zero-copy on APUs by default, but not enable it by default on discrete GPUs.
This logic follows an observation that OpenMP applications on discrete GPUs are *typically* optimized for data movement, and that configuration performs better than driver-level paging.
HIP doesn't have a concept of automatic zero-copy, as far as I know.

Reply:
Well maybe this function should be renamed then? Because I saw that the linked code above used that method to determine `is_apu_node`, which seemed identical to what this is trying to determine. So this profile is set based off of `is_apu_node` in the above source. So is this referring to something different?

Reply:
The patch (or libomptarget after patching it) doesn't have a "is_apu_node" name in it (either function or variable). I wonder if you are getting confused by successive amend's or if you are referring to downstream?

Reply:
I'm just confused, because the name suggests that the full profile implies we have USM, which is what's necessary for this behavior. The ROCR code in question uses the same logic to set a variable called `isAPU` which suggests the same logic t his `checkIfAPU` is trying to accomplish. I guess the problem is that this is more general than what we're going for here and merely want it for this single MI300 architecture?

Reply:
Leaving a comment here about investigations into rocr behavior: that code does not apply to MI300A. A new API to be able to detect an APU might be available soon, but in the meantime we need to go with the code that we have. I will modify to use the relevant ROCr API if and when it becomes available.

Reply:
Alright, maybe put a TODO here to update this function once said feature is available or something.